### PR TITLE
JSIEVE-108 Discard instruction should emmit a ActionDiscard

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

File: core/src/test/java/org/apache/jsieve/DiscardTest.java [hunk 1]
```diff
@@ -19,72 +19,49 @@
...
- */
 public class DiscardTest {
 
-    /**
-     * Test for Command 'discard' with invalid arguments
-     */
-    @org.junit.Test
-    public void testInvalidArguments() {
-        boolean isTestPassed = false;
+    private static final Condition<Action> INSTANCE_OF_ACTION_DISCARDED = new Condition<Action>() {
```
Comment (mbaechler) [inline core/src/test/java/org/apache/jsieve/DiscardTest.java:34]: `Predicates.InstanceOf`

Comment (chibenwa) [reply core/src/test/java/org/apache/jsieve/DiscardTest.java:34]: Sorry java-6 over here, then functionnal interface inference is broken here ....

Comment (mbaechler) [reply core/src/test/java/org/apache/jsieve/DiscardTest.java:34]: I can't parse your answer, what do you mean ? guava predicates is java 6, isn't it ?

File: core/src/main/java/org/apache/jsieve/tests/Header.java [hunk 2]
```diff
@@ -199,10 +202,10 @@ protected boolean match(String comparator, String matchType,
             }
         // Iterate over the header values looking for a match
         boolean isMatched = false;
-        Iterator headerValuesIter = headerValues.iterator();
+        Iterator<String> headerValuesIter = headerValues.iterator();
         while (!isMatched && headerValuesIter.hasNext()) {
```
Comment (strokyl) [inline core/src/main/java/org/apache/jsieve/tests/Header.java]: I know peple don't like it but I think a for with a return in it when you find the result is way more consise

File: core/src/test/java/org/apache/jsieve/HeaderTest.java [hunk 3]
```diff
@@ -57,6 +57,21 @@ public void testHeaderIsTrue() {
     /**
      * Test for Test 'header'
      */
+    @Test(expected = ThrowTestException.TestException.class)
+    public void testFoldedEncodedHeader() throws Exception {
```
Comment (strokyl) [inline core/src/test/java/org/apache/jsieve/HeaderTest.java]: Why don't you use expected rule ? I think the name of the test does not rely speak ? Why is it throwing an exception ?

Comment (chibenwa) [inline core/src/test/java/org/apache/jsieve/HeaderTest.java]: I don't really want to change this, it's a personal contribution...

If I sanitise these tests (which are ugly) I'll do it as part of my working time, with agreement from @rouazana  or discard the PR.

Comment (chibenwa) [inline core/src/test/java/org/apache/jsieve/HeaderTest.java]: @strokyl  sorry I deleted your comment by mistake.

I'll rename this test even if it makes it inconsistent with the other names in the class.

I'll also use ExpectedException

The exception trown is specified in the Sieve script, and it's the common mechanism used in JSieve to test for conditions. I won't change this.

File: core/src/main/java/org/apache/jsieve/tests/Header.java (hunk 2 above)
Comment (chibenwa) [reply core/src/main/java/org/apache/jsieve/tests/Header.java]: BTW you are commenting outdated commit of an other pull request. Sorry. https://github.com/linagora/james-jsieve/pull/9

File: core/src/test/java/org/apache/jsieve/HeaderTest.java (hunk 3 above)
Comment (chibenwa) [reply core/src/test/java/org/apache/jsieve/HeaderTest.java]: BTW you are commenting outdated commit of an other pull request. Sorry. https://github.com/linagora/james-jsieve/pull/9

**Comment by chibenwa** [comment]
Rebase just done on master

File: core/src/test/java/org/apache/jsieve/DiscardTest.java (hunk 1 above)
Comment (chibenwa) [reply core/src/test/java/org/apache/jsieve/DiscardTest.java:34]: The code require a Condition, you ask me to use a predicate. Without java8 and functional interfaces to bidge the gap, you a a type mismatch. Thats what I mean

Comment (mbaechler) [reply core/src/test/java/org/apache/jsieve/DiscardTest.java:34]: oh, sorry, I don't know why I thought it was a Predicate

File: core/src/test/java/org/apache/jsieve/DiscardTest.java [hunk 4]
```diff
@@ -19,72 +19,49 @@
...
+        @Override
+        public boolean matches(Action action) {
+            return action instanceof ActionDiscard;
+        }
+    };
+
+    @Test(expected = SyntaxException.class)
```
Comment (quynhn) [inline core/src/test/java/org/apache/jsieve/DiscardTest.java:41]: as usual we use ExpectedException rule. Or it just on james?

**Review by quynhn — APPROVED** [review]
Just a small concern

File: core/src/test/java/org/apache/jsieve/DiscardTest.java (hunk 4 above)
Comment (chibenwa) [reply core/src/test/java/org/apache/jsieve/DiscardTest.java:41]: All the other JSIEVE tests uses this.

**Review by rouazana — APPROVED** [review]

**Review by strokyl — APPROVED** [review]

**Pull request merged**